### PR TITLE
daml-lf fix and cleanup Decimal.fromString

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
* Fix the fonction Decimal.fromString that converts string to decimal. 
* Add tests

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Decimal.scala [hunk 1]
```diff
@@ -72,14 +72,14 @@ object Decimal {
             .setScale(scale))
     })
 
+  private val hasExpectedFormat: String => Boolean =
+    """[+-]?\d{1,28}(\.\d{1,10})?""".r.pattern.matcher(_).matches()
```
Comment (S11001001) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Decimal.scala]: Unfortunately if you don't do this the whole expression goes in the lambda body:

```suggestion
    {
      val p = """[+-]?\d{1,28}(\.\d{1,10})?""".r.pattern
      p.matcher(_).matches()
    }
```

```scala
scala> reify({
     |       val p = """[+-]?\d{1,28}(\.\d{1,10})?""".r.pattern
     |       p.matcher(_).matches()
     | }: String => Boolean)
res1: reflect.runtime.universe.Expr[String => Boolean] =
Expr[String => Boolean](({
  val p = Predef.augmentString("[+-]?\\d{1,28}(\\.\\d{1,10})?").r.pattern;
  ((x$1) => p.matcher(x$1).matches())
}: Function1[Predef.String, Boolean]))

reify(("""[+-]?\d{1,28}(\.\d{1,10})?""".r.pattern.matcher(_).matches()): String => Boolean)
res5: reflect.runtime.universe.Expr[String => Boolean] = Expr[String => Boolean]((((x$1) => Predef.augmentString("[+-]?\\d{1,28}(\\.\\d{1,10})?").r.pattern.matcher(x$1).matches()): Function1[Predef.String, Boolean]))
```

File: daml-lf/spec/daml-lf-1.rst [hunk 2]
```diff
@@ -285,7 +285,7 @@ We can now define all the literals that a program can handle::
         LitInt64  ∈ (-?)[0-9]+                      -- LitInt64:
 
   Decimal literals:
-      LitDecimal  ∈  (-?)[0-9]+.[0-9]*              -- LitDecimal
+      LitDecimal  ∈  ([-/+]?)\d{1,28}(.[0-9]\d{1-10})?  -- LitDecimal
```
Comment (S11001001) [inline daml-lf/spec/daml-lf-1.rst]:
```suggestion
      LitDecimal  ∈  ([+-]?)\d{1,28}(.[0-9]\d{1-10})?  -- LitDecimal
```

File: daml-lf/spec/value.rst [hunk 3]
```diff
@@ -253,7 +253,7 @@ digits of decimal precision.  In other words, in base-10, a number with
 A leading sign, + or -, may be optionally included.  In regular
 expression terms::
 
-  [+-]?[0-9]{0,28}(\.[0-9]{0,10})?
+  [+-]?[0-9]{1,28}(\.[0-9]{1,10})?
```
Comment (S11001001) [inline daml-lf/spec/value.rst]: Should this be considered retroactively applicable to value version 1?

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Decimal.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Decimal.scala]: Thx.

File: daml-lf/spec/daml-lf-1.rst (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/spec/daml-lf-1.rst]: Thx

File: daml-lf/spec/value.rst (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/spec/value.rst]: The current spec is clearly buggy.
For instance it was accepted ``"."`` (the string containing only dot) and ``""`` (empty string). 
Therefore we have to "retroactively" fix the spec. 

I took the intersection between 

- the originally specified reegex ```[+-]?[0-9]{0,28}(\.[0-9]{0,10})?```
- and the regexp that was actually used in the code ```[+-]?\d+(\.\d+)?```

Like that we actually do not accept less correct string representation.

Comment (bitonic) [reply daml-lf/spec/value.rst]: yes, me and @remyhaemmerle-da discussed this and we're retroactively applying this as a bug fix.

File: ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/platform/participant/util/ApiToLfEngineSpec.scala [hunk 4]
```diff
@@ -151,13 +151,13 @@ class ApiToLfEngineSpec extends WordSpec with Matchers {
     }
 
     "handle Decimals exceeding scale correctly" in {
```
Comment (bitonic) [inline ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/platform/participant/util/ApiToLfEngineSpec.scala]: now we aren't testing `ApiToLfEngine` anymore -- wrap the decimal in a value and test them through `ApiToLfEngine`

**Pull request merged**